// @framer/steveruizok.react-select/code/Select.tsx
import * as React from "react"
import {
    Frame,
    PropertyControls,
    ControlType,
    addPropertyControls,
} from "framer"
import ReactSelect from "react-select"
// import makeAnimated from "react-select/lib/animated"

export const Select = props => {
    const { cOptions, cDefaultValue, height, width, size, ...rest } = props

    if (cOptions) {
        rest.options = cOptions.map(option => {
            return {
                value: option,
                label: option,
            }
        })

        rest.defaultValue =
            rest.options.find(o => o.value == cDefaultValue) || null

        rest.value =
            typeof props.cValue === "string"
                ? rest.options.find(o => o.value == cDefaultValue)
                : props.cValue
    }

    return (
        <Frame {...{ height, width, size }} background="none">
            <ReactSelect {...rest} />
        </Frame>
    )
}

Select.defaultProps = {
    height: 200,
    width: 200,
    cOptions: ["London", "Paris", "Hong Kong", "New York"],
    cInitial: "London",
    isMulti: false,
    isClearable: false,
    styles: undefined,
}

addPropertyControls(Select, {
    cOptions: {
        title: "Options",
        type: ControlType.Array,
        propertyControl: {
            type: ControlType.String,
        },
        defaultValue: ["London", "Paris", "Hong Kong", "New York"],
    },
    cDefaultValue: {
        title: "Default",
        type: ControlType.String,
        defaultValue: "London",
    },
    isMulti: {
        title: "Multi",
        type: ControlType.Boolean,
        defaultValue: false,
    },
    isClearable: {
        title: "Clearable",
        type: ControlType.Boolean,
        defaultValue: false,
    },
    isDisabled: {
        title: "Disabled",
        type: ControlType.Boolean,
        defaultValue: false,
    },
    isSearchable: {
        title: "Searchable",
        type: ControlType.Boolean,
        defaultValue: false,
    },
})
